File: VerseReferenceImpl/src/Language/Verse/Token.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}
module Language.Verse.Token
  ( StringDelimiter(..), Token (..)
  ) where

import Language.Verse.Name

import Prettyprinter
import Prelude (Char, String, Double, Integer, Rational, fromRational)

import Text.Show

data StringDelimiter
  = Quote
  | Brace
  deriving Show

data Token
  = All
  | Ampersand
  | And
  | Array
  | At
  | AtSign
  | Block
  | Caret
  | Catch
  | Char {-# UNPACK #-} !Char
  | Class
  | Colon
  | ColonEOL
  | ColonEqual
  | ColonRightParen
  | Comma
  | Dedent
  | Divide
  | DivideEqual
  | Do
  | Dot
  | DotDot
  | DotSpace
  | EOF
  | Else
  | Enum
  | Equal
  | Exists
  | Fail
  | Fails
  | False
  | FatArrow
  | Float !Rational
  | For
  | Forall
  | Function
  | Greater
  | GreaterEqual
  | If
  | Indent
  | Int !Integer
  | LeftBrace
  | LeftBracket
  | LeftParen
  | Less
  | LessEqual
  | Minus
  | MinusEqual
  | Module
  | Multiply
  | MultiplyEqual
  | Name Name
  | Path Name
  | Prefix Name
  | Newline
  | Not
  | NotEqual
  | Of
  | One
  | Option
  | Or
  | Pipe
  | Plus
  | PlusEqual
  | QuestionMark
  | Return
  | RightBrace
  | RightBracket
  | RightParen
  | Semi
  | Semi'  -- Inserted by Indenter instead of newline in some case, in enum shoudl be treated as a comma
  | Set
  | String StringDelimiter String StringDelimiter
  | Struct
  | Sync
  | Then
  | ThinArrow
  | Tilde
  | True
  | Truth
  | Until
  | Var
  | Where deriving Show

instance Pretty Token where
  pretty = \ case
    All -> "all"
    Ampersand -> "&"
    And -> "and"
    Array -> "array"
    At -> "at"
    AtSign -> "@"
    Block -> "block"
    Caret -> "^"
    Catch -> "catch"
    Char x -> pretty ['\'', x, '\'']
    Class -> "class"
    Colon -> colon
    ColonEOL -> colon
    ColonEqual -> colon <> equals
    ColonRightParen -> colon <> rparen
    Comma -> comma
    Dedent -> "dedent"
    Divide -> pretty '/'
    DivideEqual -> "/="
    Do -> "do"
    Dot -> dot
    DotDot -> ".."
    DotSpace -> "."
    EOF -> "end" <+> "of" <+> "file"
    Else -> "else"
    Enum -> "enum"
    Equal -> equals
    Exists -> "exists"
    Fail -> "fail"
    Fails -> "fails"
    False -> "false"
    FatArrow -> equals <> rangle
    Float x -> pretty (fromRational x :: Double)
    For -> "for"
    Forall -> "forall"
    Function -> "function"
    Greater -> pretty '>'
    GreaterEqual -> ">="
    If -> "if"
    Indent -> "indent"
    Int x -> pretty x
    LeftBrace -> lbrace
    LeftBracket -> lbracket
    LeftParen -> lparen
    Less -> pretty '<'
    LessEqual -> "<="
    Minus -> pretty '-'
    MinusEqual -> "-="
    Module -> "module"
    Multiply -> pretty '*'
    MultiplyEqual -> "*="
    Name x -> pretty x
    Path x -> pretty x
    Prefix x -> "'" <> pretty x <> "'"
    Newline -> "newline"
    Not -> "not"
    NotEqual -> "<>"
    Of -> "of"
    One -> "one"
    Option -> "option"
    Or -> "or"
    Pipe -> pipe
    Plus -> pretty '+'
    PlusEqual -> "+="
    QuestionMark -> pretty '?'
    Return -> "return"
    RightBrace -> rbrace
    RightBracket -> rbracket
    RightParen -> rparen
    Semi -> semi
    Semi' -> semi<>"'"
    Set -> "set"
    String begin x end -> prettyBegin begin <> pretty x <> prettyEnd end
    Struct -> "struct"
    Sync -> "sync"
    Then -> "then"
    ThinArrow -> pretty '-' <> rangle
    Tilde -> "~"
    True -> "true"
    Truth -> "truth"
    Until -> "until"
    Var -> "var"
    Where -> "where"
    where
      prettyBegin Quote = "\""
      prettyBegin Brace = "}"

      prettyEnd Quote = "\""
      prettyEnd Brace = "{"

File: Tsx/src/parser/token.ts
export enum StringDelimiter {
  Quote = 'Quote',
  Brace = 'Brace'
}

export type Token =
  | { kind: 'All' }
  | { kind: 'Ampersand' }
  | { kind: 'And' }
  | { kind: 'Array' }
  | { kind: 'At' }
  | { kind: 'AtSign' }
  | { kind: 'Block' }
  | { kind: 'Caret' }
  | { kind: 'Catch' }
  | { kind: 'Char'; value: string }
  | { kind: 'Class' }
  | { kind: 'Colon' }
  | { kind: 'ColonEOL' }
  | { kind: 'ColonEqual' }
  | { kind: 'ColonRightParen' }
  | { kind: 'Comma' }
  | { kind: 'Dedent' }
  | { kind: 'Divide' }
  | { kind: 'DivideEqual' }
  | { kind: 'Do' }
  | { kind: 'Dot' }
  | { kind: 'DotDot' }
  | { kind: 'DotSpace' }
  | { kind: 'EOF' }
  | { kind: 'Else' }
  | { kind: 'Enum' }
  | { kind: 'Equal' }
  | { kind: 'Exists' }
  | { kind: 'Fail' }
  | { kind: 'Fails' }
  | { kind: 'False' }
  | { kind: 'FatArrow' }
  | { kind: 'Float'; value: number }
  | { kind: 'For' }
  | { kind: 'Forall' }
  | { kind: 'Function' }
  | { kind: 'Greater' }
  | { kind: 'GreaterEqual' }
  | { kind: 'If' }
  | { kind: 'Indent' }
  | { kind: 'Int'; value: bigint }
  | { kind: 'LeftBrace' }
  | { kind: 'LeftBracket' }
  | { kind: 'LeftParen' }
  | { kind: 'Less' }
  | { kind: 'LessEqual' }
  | { kind: 'Minus' }
  | { kind: 'MinusEqual' }
  | { kind: 'Module' }
  | { kind: 'Multiply' }
  | { kind: 'MultiplyEqual' }
  | { kind: 'Name'; value: string }
  | { kind: 'Path'; value: string }
  | { kind: 'Prefix'; value: string }
  | { kind: 'Newline' }
  | { kind: 'Not' }
  | { kind: 'NotEqual' }
  | { kind: 'Of' }
  | { kind: 'One' }
  | { kind: 'Option' }
  | { kind: 'Or' }
  | { kind: 'Pipe' }
  | { kind: 'Plus' }
  | { kind: 'PlusEqual' }
  | { kind: 'QuestionMark' }
  | { kind: 'Return' }
  | { kind: 'RightBrace' }
  | { kind: 'RightBracket' }
  | { kind: 'RightParen' }
  | { kind: 'Semi' }
  | { kind: 'SemiPrime' } // Semi' in Haskell
  | { kind: 'Set' }
  | { kind: 'String'; begin: StringDelimiter; value: string; end: StringDelimiter }
  | { kind: 'Struct' }
  | { kind: 'Sync' }
  | { kind: 'Then' }
  | { kind: 'ThinArrow' }
  | { kind: 'Tilde' }
  | { kind: 'True' }
  | { kind: 'Truth' }
  | { kind: 'Until' }
  | { kind: 'Var' }
  | { kind: 'Where' }
  | { kind: 'Break' }
  | { kind: 'Continue' }
  | { kind: 'Yield' }
  | { kind: 'Next' }
  | { kind: 'Over' }
  | { kind: 'While' }
  | { kind: 'When' }
  // Specifiers
  | { kind: 'Decides' }
  | { kind: 'Succeeds' }
  | { kind: 'Fails' }
  | { kind: 'Transacts' }
  | { kind: 'Computes' }
  | { kind: 'Ambiguates' }
  | { kind: 'Reads' }
  | { kind: 'Writes' }
  | { kind: 'Allocates' }
  | { kind: 'Suspends' };

// Helper functions for creating tokens
export function createIntToken(value: bigint): Token {
  return { kind: 'Int', value };
}

export function createFloatToken(value: number): Token {
  return { kind: 'Float', value };
}

export function createCharToken(value: string): Token {
  return { kind: 'Char', value };
}

export function createNameToken(value: string): Token {
  return { kind: 'Name', value };
}

export function createStringToken(begin: StringDelimiter, value: string, end: StringDelimiter): Token {
  return { kind: 'String', begin, value, end };
}

export function createPathToken(value: string): Token {
  return { kind: 'Path', value };
}

export function createPrefixToken(value: string): Token {
  return { kind: 'Prefix', value };
}